File: Algorithms_Part_1/Algorithm1Sample/Ex1QuickFind.java
/* *****************************************************************************
 *  Name:
 *  Date:
 *  Description:
 **************************************************************************** */

import edu.princeton.cs.algs4.StdIn;
import edu.princeton.cs.algs4.StdOut;

public class Ex1QuickFind {

    private int[] data;

    public Ex1QuickFind(int n) {
        data = new int[n];
        for (int i = 0; i < n; i++)
            data[i] = i;
    }

    public void union(int p, int q) {
        int pid = data[p];
        int qid = data[q];
        for (int i = 0; i < data.length; i++) {
            if (data[i] == pid)
                data[i] = qid;
        }
    }

    public boolean isConnected(int p, int q) {
        return data[p] == data[q];
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();
        for (int i = 0; i < data.length; i++)
            result.append(data[i]).append(" ");
        return result.toString();
    }

    public static void main(String[] args) {

        int n = StdIn.readInt();
        Ex1QuickFind qf = new Ex1QuickFind(n);
        for (int i = 0; i < n; i++) {
            int p = StdIn.readInt();
            int q = StdIn.readInt();

            if (!qf.isConnected(p, q)) {
                qf.union(p, q);
                StdOut.println("Connected(" + p + "," + q + ") : " + qf.toString());
            }
        }

    }
}

File: PythonLearnig/Algorithms_Part_1/Ex1QuickFind.py
class QuickFind:

    def __init__(self, n):
        self.data = list(range(n))

    def union(self, p, q):
        pid = self.data[p]
        qid = self.data[q]
        for i in range(len(self.data)):
            if self.data[i] == pid:
                self.data[i] = qid

    def is_connected(self, p, q):
        return self.data[p] == self.data[q]

    def __str__(self):
        return str(self.data)


def test_quick_find():

    n = int(input("Please enter the no of vertices: "))
    qf = QuickFind(n)

    for i in range(n):
        p = int(input())
        q = int(input())

        if not qf.is_connected(p,q):
            qf.union(p,q)
            print(f"Connected({p},{q}) : {qf}")


if __name__ == "__main__":
    test_quick_find()
